### PR TITLE
[red-knot] Type narrow in else clause

### PR DESCRIPTION
**Pull request by TomerBin**
Added support for type narrowing in elif and else scopes as part of #13694.

## Summary

This is ready for review, though still somewhat of a draft. Please let me know if I addressed this correctly or if a completely different solution would be better :)

## Test Plan

- mdtest
- builder unit test for union negation.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types/builder.rs [hunk 1]
```diff
@@ -667,6 +663,27 @@ mod tests {
...
+        assert_eq!(union.display(&db).to_string(), "Literal[1, 2]");
+
+        let ty = IntersectionBuilder::new(&db).add_negative(union).build();
+
+        let expected = IntersectionBuilder::new(&db)
+            .add_negative(Type::IntLiteral(1))
+            .add_negative(Type::IntLiteral(2))
+            .build();
+
+        assert_eq!(ty.display(&db).to_string(), "~Literal[1] & ~Literal[2]");
```
Comment (TomerBin) [inline crates/red_knot_python_semantic/src/types/builder.rs:683]: I didn't find any another builder tests that assert the `display` value of the result type but I actually find it more expressive and readable that way. LMK if I should better stick to the other tests' style.

File: crates/red_knot_python_semantic/src/semantic_index/constraint.rs [hunk 2]
```diff
@@ -7,7 +7,13 @@ use crate::semantic_index::expression::Expression;
...
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub(crate) enum Constraint<'db> {
+pub(crate) struct Constraint<'db> {
+    pub(crate) node: ConstraintNode<'db>,
+    pub(crate) negative: bool,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub(crate) enum ConstraintNode<'db> {
```
Comment (TomerBin) [inline crates/red_knot_python_semantic/src/semantic_index/constraint.rs:16]: In this context, "constraint" term is used for both the node that applies the constraint, and for the outcome of the narrowing process. 
What do you think about renaming this one into `TypeGuard` and use `constraint` name only for the output (map from symbol ids to types)?

Comment (carljm) [reply crates/red_knot_python_semantic/src/semantic_index/constraint.rs:16]: Hmm. I'm open to a renaming here to better distinguish these, but not sure I like `TypeGuard`, I'd rather avoid that term outside of dealing with the actual `typing.TypeGuard` feature.

File: crates/red_knot_python_semantic/src/types/narrow.rs [hunk 3]
```diff
@@ -51,15 +51,30 @@ fn all_narrowing_constraints_for_pattern<'db>(
...
+        },
+    )
+    .finish()
 }
 
 #[salsa::tracked(return_ref)]
 fn all_narrowing_constraints_for_expression<'db>(
     db: &'db dyn Db,
     expression: Expression<'db>,
+    negative: bool,
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/narrow.rs:68]: Hmm, I think Salsa does something kind of inefficient when a tracked function has multiple inputs; it internally interns all the arguments. Although it's slightly weird-looking, I think we will get more efficient results here if we break into two separate Salsa tracked functions, `all_narrowing_constraints_for_expression` and `all_negative_narrowing_constraints_for_expression`.

File: crates/red_knot_python_semantic/src/types/narrow.rs [hunk 4]
```diff
@@ -51,15 +51,30 @@ fn all_narrowing_constraints_for_pattern<'db>(
...
 ) -> NarrowingConstraints<'db> {
-    NarrowingConstraintsBuilder::new(db, Constraint::Expression(expression)).finish()
+    NarrowingConstraintsBuilder::new(
+        db,
+        Constraint {
+            node: ConstraintNode::Expression(expression),
+            negative,
+        },
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/narrow.rs:75]: Is it worth re-building a `Constraint` struct here vs just passing in the two elements separately to the builder? It mostly seems like it just results in more verbose usage inside the builder (`self.constraint.negative` vs `self.negative`).

File: crates/red_knot_python_semantic/src/types/narrow.rs [hunk 5]
```diff
@@ -192,29 +220,42 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
             {
                 // SAFETY: we should always have a symbol for every Name node.
                 let symbol = self.symbols().symbol_id_by_name(id).unwrap();
-                let comp_ty = inference.expression_ty(right.scoped_ast_id(self.db, scope));
-                match op {
-                    ast::CmpOp::IsNot => {
-                        if comp_ty.is_singleton() {
+                let rhs_ty = inference.expression_ty(right.scoped_ast_id(self.db, scope));
+                match (op, negative) {
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/narrow.rs:224]: We could make the match arms a bit less repetitive/verbose here (though arguably also less explicit?) if we call a function here that returns the negated `ast::CmpOp` if `negative` and then match on the resulting `CmpOp`. What do you think?

File: crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_elif_else.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+## Positive contributions become negative in elif-else blocks
+
+```py
+def int_instance() -> int: ...
+
+x = int_instance()
+
+if x == 1:
+    reveal_type(x)  # revealed: Literal[1]
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_elif_else.md:11]: We've previously discussed that we can't actually safely do this narrowing, because `x` here could be an instance of some subclass of `int` that overrides `__eq__` to compare equal to `1`.

Perhaps experience will tell us that we are being too conservative here, and we should just ignore that possibility; but I'd rather not do that until/unless it's clear that we need to.

File: crates/red_knot_python_semantic/src/types/narrow.rs [hunk 7]
```diff
@@ -192,29 +220,42 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
...
+                        } else if let Some(union) = rhs_ty.into_union_type() {
+                            if union
+                                .elements(self.db)
+                                .iter()
+                                .all(|ty| ty.is_single_valued(self.db))
+                            {
+                                self.constraints.insert(symbol, rhs_ty);
+                            }
+                        }
+                    }
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/narrow.rs:258]: As discussed in another comment in the tests, I don't think we can generally narrow on equality checks (or negative inequality checks). But I think we can (and probably will need to) do a more limited form of this, where we check the type of the LHS and if it is a union of single-valued types, we eliminate (by constraining with a negative intersection of) any single-valued types that we know are disjoint with the RHS.

In other words, while we can't infer that `x: int; if x == 1:` results in `Literal[1]`, we can infer that `x: Literal[1, 2, 3]; if x == 1:` results in `Literal[1]`, and I think we'll need to be able to make the latter inference.

(There may be a more general way to frame this than what I wrote above, need to think about it.)

**Review by carljm — COMMENTED** [review]
This is really excellent, thank you for doing this!!

Some comments below; maybe don't put too much work into addressing comments until @AlexWaygood and/or @sharkdp have a chance to take a look.

File: crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is.md [hunk 8]
```diff
@@ -37,4 +41,20 @@ reveal_type(y)  # revealed: bool
 if y is x is False:  # Interpreted as `(y is x) and (x is False)`
     reveal_type(x)  # revealed: Literal[False]
     reveal_type(y)  # revealed: bool
+else:
+    reveal_type(x)  # revealed: bool
```
[the remark this answers is not in the document]
Comment (sharkdp) [reply crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is.md:45]: No, because there are two conditions in the `if` clause. And only one of them could be `False`: if `y = True` and `x = False`, we also end up in this branch.

File: crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is_not.md [hunk 9]
```diff
@@ -51,4 +57,7 @@ reveal_type(y)  # revealed: bool
 if y is not x is not False:  # Interpreted as `(y is not x) and (x is not False)`
     reveal_type(x)  # revealed: Literal[True]
     reveal_type(y)  # revealed: bool
+else:
+    reveal_type(x)  # revealed: bool
```
[the remark this answers is not in the document]
Comment (sharkdp) [reply crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is_not.md:61]: No. Similar to above: if `x = y = False`, we also end up in this branch.

File: crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is_not.md
Comment (sharkdp) [inline crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is_not.md:1]: In this file, can we maybe also add something equivalent to the test "`is` for other types" test, but for `is not`:
```py
class A: ...

x = A()
y = x if flag else None

if y is not x:
    reveal_type(y)  # revealed: A | None
else:
    reveal_type(y)  # revealed: A
```

This wouldn't have been very interesting/useful before. But now with the `else` branch, it is.

File: crates/red_knot_python_semantic/src/semantic_index/constraint.rs (hunk 2 above)
Comment (MichaReiser) [reply crates/red_knot_python_semantic/src/semantic_index/constraint.rs:16]: We often called these types *Kind* in other places.

File: crates/red_knot_python_semantic/resources/mdtest/narrow/isinstance.md [hunk 10]
```diff
@@ -76,6 +83,15 @@ if isinstance(x, A):
...
         reveal_type(x)  # revealed: A & B
+    else:
+        reveal_type(x)  # revealed: A & ~B
+
+if isinstance(x, (A, B)):
+    reveal_type(x)  # revealed: A | B
+elif isinstance(x, (A, C)):
+    reveal_type(x)  # revealed: C & ~A & ~B
+else:
+    reveal_type(x)  # revealed: object & ~A & ~B & ~C
```
Comment (sharkdp) [inline crates/red_knot_python_semantic/resources/mdtest/narrow/isinstance.md:94]: Not sure if always desirable, but we could maybe add a `TODO` here to always remove `object` from the positive side of an intersection.

File: crates/red_knot_python_semantic/src/semantic_index/constraint.rs (hunk 2 above)
Comment (TomerBin) [reply crates/red_knot_python_semantic/src/semantic_index/constraint.rs:16]: What do you think about "Predicate"? I like that it's more easy to understand what "negative" means on a predicate, than on Type or Kind as of the negation of a predicate doesn't always imply the negation of its outcome constraint.

File: crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (carljm) [reply crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_is.md:45]: Unresolving this conversation because it might be useful to add a comment here explaining this?

File: crates/red_knot_python_semantic/src/semantic_index/constraint.rs (hunk 2 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/semantic_index/constraint.rs]: I'm happy with "Predicate"!

File: crates/red_knot_python_semantic/src/types/narrow.rs (hunk 7 above)
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/narrow.rs]: This represents a more general category of narrowing cases that we will also run into with, for example, simple `if not x:` narrowing, where the information gained from the narrowing in that case is "`x` is falsy" and in this case is "`x` is equal to 1", neither of which is a constraint that can correctly be expressed simply a type to intersect `x` with, given our currently available types. I spent some more time thinking about this, and wrote it up at https://github.com/astral-sh/ruff/issues/13694#issuecomment-2438438759

I'm afraid that's a bit long, but I think the relevant conclusion for this PR is, let's not handle it in this PR :) For this PR, let's just remove the narrowing on `==` (or negated `!=`) entirely for now.

**Comment by carljm** [comment]
@TomerBin are you ready for another review of this? There are still some previous comments unresolved.

**Comment by TomerBin** [comment]
> @TomerBin are you ready for another review of this? There are still some previous comments unresolved.

@carljm Not yet.. I wanted to ask something regarding __eq__. The logic prior to this PR does add single-value types in the rhs as a negative constraint on the symbol, without considering whether the lhs is single-valued.
```py
ast::CmpOp::NotEq => {
                        if rhs_ty.is_single_valued(self.db) {
                            let ty = IntersectionBuilder::new(self.db)
                                .add_negative(rhs_ty)
                                .build();
                            self.constraints.insert(symbol, ty);
                        }
                    }
```
In the scope of this PR - do you think this logic should be adopted as well? Or should I just ignore the negation of this predicate at all?

Thanks :)

**Comment by carljm** [comment]
Although it's not safe to apply an intersection with `Literal[1]` as a result of `if x == 1`, it is safe to apply an intersection with `~Literal[1]` as a result of `if x != 1`. In the former case, we can't say that `x` must be `Literal[1]`, it could be some other type that just compares equal to `1`. But in the latter case, we can definitively say that `x` cannot be `Literal[1]`, because we know that ever inhabitant of `Literal[1]` must compare equal to `1`.

So the current code is correct.

And in this PR, I think it is always the case that if we can infer something from `x != y`, then we can infer the same thing from the `else` clause of `if x == y`.

So yes, I think we should, in this PR, infer that in the `else` clause of `if x == 1`, we apply a constraint of `~Literal[1]` to `x`.

File: crates/red_knot_python_semantic/src/types/builder.rs [hunk 11]
```diff
@@ -667,6 +663,27 @@ mod tests {
         assert_eq!(ty, Type::IntLiteral(1));
     }
 
+    #[test]
+    fn build_negative_union_de_morgan() {
```
[the remark this answers is not in the document]
Comment (TomerBin) [reply crates/red_knot_python_semantic/src/types/builder.rs:667]: Im not sure how to test that in a unit test. I need to generate an intersection between two types that won't be simplified (so they can't be disjoint). If it was an mdtest I'd do it with class literals, but I'm not sure how to that that in here. Do you have an idea?
Maybe Ill prefer to do it in separate PR if you're happy with that.

[the remark this answers is not in the document]
Comment (carljm) [reply crates/red_knot_python_semantic/src/types/builder.rs:667]: Yes, separate PR is totally fine! I've created an issue just to make sure we don't lose track of it, in case you don't have time to do it: https://github.com/astral-sh/ruff/issues/13931

I think you can use known types from typeshed to create the intersection? E.g. `KnownClass::Str.to_instance()` and `KnownClass::Int.to_instance()` can be intersected without simplifying.

(Technically these cannot be intersected because they are both builtin types with different memory layouts, so you can't subclass both of them. But we don't represent this yet.)

I think you can also just `db.write_file(...)` a file with a couple class definitions in it, much like the tests in `infer.rs`, and then pull out those class types. It's a bit more boilerplate, and there aren't existing tests in here that do it, but it's fine to do it.

**Comment by TomerBin** [comment]
I think that It's ready for another CR :)

File: crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_elif_else.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
+# Narrowing for conditionals with elif and else
+
+## Positive contributions become negative in elif-else blocks
+
+```py
+def int_instance() -> int: ...
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_elif_else.md:6]: We should fill in the bodies of helper functions in tests so that they don't start failing in the future when we start checking that functions actually return what they say they return:

```suggestion
def int_instance() -> int:
    return 42
```

File: crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_not_eq.md [hunk 13]
```diff
@@ -52,3 +74,17 @@ y = int_instance()
 if x != y:
     reveal_type(x)  # revealed: int | None
 ```
+
+## Mix of single-valued and non-single-valued types
+
+```py
+def int_instance() -> int: ...
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/resources/mdtest/narrow/conditionals_not_eq.md:81]:
```suggestion
def int_instance() -> int:
    return 42
```

File: crates/red_knot_python_semantic/src/semantic_index/constraint.rs [hunk 14]
```diff
@@ -7,13 +7,19 @@ use crate::semantic_index::expression::Expression;
 use crate::semantic_index::symbol::{FileScopeId, ScopeId};
 
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub(crate) enum Constraint<'db> {
+pub(crate) struct Predicate<'db> {
+    pub(crate) node: PredicateNode<'db>,
+    pub(crate) negative: bool,
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/semantic_index/constraint.rs:12]: It feels slightly odd to me that a negative boolean value for this field indicates a positive predicate. Could we have the meaning flipped, so that a positive value for the field indicates a positive predicate (and vice versa)?

```suggestion
    pub(crate) is_positive: bool,
```

**Review by AlexWaygood — COMMENTED** [review]
(Just a few comments from skimming, haven't looked properly yet!)

Looks like some of your tests are failing following https://github.com/astral-sh/ruff/commit/6aaf1d944657a7f37bd99c781c2b1ebdebfdf0e3 being pushed to `main` — you'll need to define the `flag` variables in your tests in the same way we did in that commit. (You may need to merge in `main` or rebase to see the same test failures locally)

**Review by carljm — APPROVED** [review]
Looks good! 90% of my comments are just that we should have TODOs in tests reflecting the equality narrowing for literal types that we should do in future, as discussed. I'll just apply these suggested changes.

It also looks to me like the constraint -> predicate rename isn't quite fully carried through. If it's OK I might just also push some changes to complete that, and then go ahead and land this.

**Comment by carljm** [comment]
After looking at it more, I think for now I'm just going to revert the Predicate name change in this PR. There's still a ton of things named "constraint" inside the use-def map and semantic index, so the current change just leaves us in an even less consistent naming situation. And I'm not convinced enough that the current naming is a problem to be motivated to fully carry the rename through.

**Comment by TomerBin** [comment]
LGTM 👍👍
